visualization/app/codeCharta/state/store/metricData/edgeMetricData/edgeMetricData.service.ts

```diff
@@ -0,0 +1,111 @@
+import { isActionOfType } from "../../../../util/reduxHelper"
+import {
+	AttributeTypes,
+	AttributeTypeValue,
+	BlacklistItem,
+	CodeMapNode,
+	EdgeMetricCount,
+	EdgeMetricData
+} from "../../../../codeCharta.model"
+import { FileState } from "../../../../model/files/files"
+import { HierarchyNode } from "d3"
+import { BlacklistService, BlacklistSubscriber } from "../../fileSettings/blacklist/blacklist.service"
+import { FilesSelectionSubscriber, FilesService } from "../../files/files.service"
+import { nodeEdgeMetricsMap } from "./edgeMetricData.reducer"
+import { AttributeTypesService, AttributeTypesSubscriber } from "../../fileSettings/attributeTypes/attributeTypes.service"
+
+export interface EdgeMetricDataSubscriber {
+	onEdgeMetricDataChanged(edgeMetricData: EdgeMetricData[])
+}
+
+export class EdgeMetricDataService implements StoreSubscriber, BlacklistSubscriber, FilesSelectionSubscriber, AttributeTypesSubscriber {
+	private static EDGE_METRIC_DATA_CHANGED_EVENT = "edge-metric-data-changed"
+	public static NONE_METRIC = "None"
+
+	constructor(private $rootScope: IRootScopeService, private storeService: StoreService) {
+		StoreService.subscribe(this.$rootScope, this)
+		BlacklistService.subscribe(this.$rootScope, this)
+		FilesService.subscribe(this.$rootScope, this)
+		AttributeTypesService.subscribe(this.$rootScope, this)
+	}
+
+	public onStoreChanged(actionType: string) {
+		if (isActionOfType(actionType, EdgeMetricDataActions)) {
+			this.notify(this.select())
+		}
+	}
+
+	public onBlacklistChanged(blacklist: BlacklistItem[]) {
+		this.storeService.dispatch(calculateNewEdgeMetricData(this.storeService.getState().files, blacklist))
+	}
+
+	public onFilesSelectionChanged(files: FileState[]) {
+		this.storeService.dispatch(calculateNewEdgeMetricData(files, this.storeService.getState().fileSettings.blacklist))
+	}
+
+	public onAttributeTypesChanged(attributeTypes: AttributeTypes) {
+		this.storeService.dispatch(
+			calculateNewEdgeMetricData(this.storeService.getState().files, this.storeService.getState().fileSettings.blacklist)
+		)
+	}
+
+	public getMetricNames(): string[] {
+		if (!this.storeService.getState().metricData.edgeMetricData) {
+			return []
+		}
+
+		return this.storeService.getState().metricData.edgeMetricData.map(x => x.name)
+	}
+
+	public getAmountOfAffectedBuildings(metricName: string): number {
+		if (!nodeEdgeMetricsMap || !nodeEdgeMetricsMap.has(metricName)) {
+			return 0
+		}
+		return nodeEdgeMetricsMap.get(metricName).size
+	}
+
+	public getNodesWithHighestValue(metricName: string, numberOfNodes: number): string[] {
+		if (!nodeEdgeMetricsMap || !nodeEdgeMetricsMap.has(metricName)) {
+			return []
+		}
+
+		const highestEdgeCountBuildings: string[] = []
+		const edgeMetricMapKeyIterator = nodeEdgeMetricsMap.get(metricName).keys()
+		for (let i = 0; i < numberOfNodes; i++) {
+			highestEdgeCountBuildings.push(edgeMetricMapKeyIterator.next().value)
+		}
+
+		return highestEdgeCountBuildings
```

Comment:
```suggestion
		const nodeEdgeMetrics = nodeEdgeMetricsMap.get(metricName)
		return nodeEdgeMetrics === undefined ? [] : [...nodeEdgeMetrics.keys()]
```

Reply:
This is not correct. Your suggestion does not take the parameter `numberOfNodes` into account. This parameter is actually `amountOfEdgePreviews`. This means: display x buildings with the highest amount of incoming and outgoing edges.


```ts
const nodeEdgeMetrics = nodeEdgeMetricsMap.get(metricName)
return nodeEdgeMetrics === undefined ? [] : [...nodeEdgeMetrics.keys()].slice(0, numberOfNodes)
```

We need to splice at the end. I'll rename the parameter to avoid confusion.

Reply:
Thanks, I missed that! Slicing is not ideal, since we would end up doing more work than before. Another alternative is the following (not saying that it's definitely better, I personally just prefer this notation over `.next().value`):

```js
const nodeEdgeMetrics = nodeEdgeMetricsMap.get(metricName)
const keys = []
if (nodeEdgeMetrics === undefined) {
  return keys
}
for (const key of nodeEdgeMetrics.keys()) {
  keys.push(key)
  if (keys.length === numberOfNodes) {
    break;
  }
}
return keys
```

Reply:
Oh, true. Slicing would require us to traverse the array again. Thanks for the suggestion